api/grpcserver/post_client.go

```diff
@@ -29,82 +29,132 @@ func newPostClient(con chan<- postCommand) *postClient {
-				},
+	req := &pb.NodeRequest{
+		Kind: &pb.NodeRequest_GenProof{
+			GenProof: &pb.GenProofRequest{
+				Challenge: challenge,
 			},
 		},
-		resp: resp,
 	}
 
+	var proofResp *pb.GenProofResponse
 	for {
-		// send command
-		select {
-		case <-pc.closed:
-			return nil, nil, fmt.Errorf("post client closed")
-		case <-ctx.Done():
-			return nil, nil, ctx.Err()
-		case pc.con <- cmd:
+		resp, err := pc.send(ctx, req)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		proofResp = resp.GetGenProof()
+		if proofResp == nil {
+			return nil, nil, fmt.Errorf("unexpected response of type: %T", resp.GetKind())
+		}
+
+		switch proofResp.GetStatus() {
+		case pb.GenProofStatus_GEN_PROOF_STATUS_ERROR:
+			return nil, nil, fmt.Errorf("error generating proof: %s", proofResp)
+		case pb.GenProofStatus_GEN_PROOF_STATUS_UNSPECIFIED:
+			return nil, nil, fmt.Errorf("unspecified error generating proof: %s", proofResp)
+		case pb.GenProofStatus_GEN_PROOF_STATUS_OK:
+		default:
+			return nil, nil, fmt.Errorf("unknown status: %s", proofResp)
+		}
+
+		if proofResp.GetProof() != nil {
+			break
 		}
 
-		// receive response
 		select {
-		case <-pc.closed:
-			return nil, nil, fmt.Errorf("post client closed")
 		case <-ctx.Done():
 			return nil, nil, ctx.Err()
-		case resp := <-resp:
-			proofResp := resp.GetGenProof()
-			if proofResp == nil {
-				return nil, nil, fmt.Errorf("unexpected response of type: %T", resp.GetKind())
-			}
-			switch proofResp.GetStatus() {
-			case pb.GenProofStatus_GEN_PROOF_STATUS_OK:
-				if proofResp.GetProof() == nil {
-					select {
-					case <-ctx.Done():
-						return nil, nil, ctx.Err()
-					case <-time.After(2 * time.Second):
-						// TODO(mafa): make polling interval configurable
-						continue
-					}
-				}
-
-				proof := proofResp.GetProof()
-				proofMeta := proofResp.GetMetadata()
-				if proofMeta == nil {
-					return nil, nil, fmt.Errorf("proof metadata is nil")
-				}
-
-				if !bytes.Equal(proofMeta.GetChallenge(), challenge) {
-					return nil, nil, fmt.Errorf("unexpected challenge: %x", proofMeta.GetChallenge())
-				}
-
-				postMeta := proofMeta.GetMeta()
-				if postMeta == nil {
-					return nil, nil, fmt.Errorf("post metadata is nil")
-				}
-
-				return &types.Post{
-						Nonce:   proof.GetNonce(),
-						Indices: proof.GetIndices(),
-						Pow:     proof.GetPow(),
-					}, &types.PostMetadata{
-						Challenge:     proofMeta.GetChallenge(),
-						LabelsPerUnit: postMeta.GetLabelsPerUnit(),
-					}, nil
-			case pb.GenProofStatus_GEN_PROOF_STATUS_ERROR:
-				return nil, nil, fmt.Errorf("error generating proof: %s", proofResp)
-			case pb.GenProofStatus_GEN_PROOF_STATUS_UNSPECIFIED:
-				return nil, nil, fmt.Errorf("unspecified error generating proof: %s", proofResp)
-			}
+		case <-time.After(2 * time.Second):
+			// TODO(mafa): make polling interval configurable
+			continue
 		}
 	}
+
+	proof := proofResp.GetProof()
+	metaData := proofResp.GetMetadata()
+	if metaData == nil {
```

Comment:
This looks weird. How about:
```suggestion
	metadata := proofResp.GetMetadata()
	if metadata == nil {
```